WikiWatershed/model-my-watershed
# HydroShare Export: Include Model Result Tables as CSV

## Overview

Exports model results for MapShed and TR-55 projects as CSVs identical to their tables to HydroShare. This is done by cycling through all scenarios in the UI, rendering the tables, exporting them, and collecting the results. The scenario switching is done in the background, and the previously user-selected scenario is reselected in the end.

The additional data gathering step, and the additional files being exported to HydroShare, delay the export process. I can see it becoming quite time taking with a handful of active scenarios. Optimizing this process is something we can look in to later as an enhancement (see #2592).

Switching through all the scenarios can potentially affect the scroll state of the sidebar.

Connects #2666 
Connects #2667 

### Demo

![2018-02-27 16 46 39](https://user-images.githubusercontent.com/1430060/36756897-51dd4b2e-1bde-11e8-8e94-8843292af1a2.gif)

### Notes

One of the interim commits 1272ac8 exports models as JSON, directly in the format returned from the API. Instead of omitting this work, I decided to leave it in for history.

I also had to upgrade the [`tableExport`](https://github.com/hhurz/tableExport.jquery.plugin) to the latest `1.9.9` to be able to export hidden tables. Existing functionality for downloads seems to be unaffected.

## Testing Instructions

* Check out this branch and `bundle --debug --vendor`
* Go to [:8000/](http://localhost:8000/) create a TR-55 project.
* Add a scenario. Make some modifications.
* Export the project to HydroShare. Inspect the export to ensure that all the model results are identical to the tables, and to the exports generated from the UI.
* Do the same for a MapShed project. Ensure that its values correspond as well.
* Ensure the export names match up with the specifications in the original issues.
* Ensure CSV export still functions as before.

This is a potentially complex code review. I will be available to pair on this if necessary.
